docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,137 @@
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority`, `sender`, and `sequence` , into the
+`ResponseCheckTx` type and augment the existing mempool data structure to
+facilitate prioritization of uncommitted transactions in addition to extended
+functionality such as replace-by-priority and allowing multiple transactions to
+exist from the same sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
++ int64          sequence   = 11;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s), and
+the `sequence` could be the signer's sequence number.
+
+Only `sender` is required, while `priority` and `sequence` can be omitted which
+would result in using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:[]Tx>`, i.e a mapping from `sender` to a list of `Tx`
+objects, where each `Tx` is ordered by monotonically increasing, but not
+necessarily consecutive, `sequence` values. For example, the sender `Alice`, whose
+current committed sequence value is `3`, could have transactions in this map with
+sequence values `<4,5,6,8>`, where only the transaction with sequence `4` could be
+processed next.
+
+Note, the mempool itself does not and will not know about the latest committed
+sequence value, but it is assumed that all transactions that end up in the mempool
+pass `CheckTx`. In addition, the application and thus Tendermint's mempool does
+not necessarily have to utilize the concept of a sequence.
+
+On top of this mapping, we index all transactions by priority using a priority
+queue. When a proposer is ready to reap transactions for the next block proposal,
+transactions are selected from this priority index. For each priority-indexed
+transaction, we evaluate if that transaction is at the head of the sender's
+list of transactions, if so, it is included in the block and the index is removed
+along with the transaction itself.
+
```

Comment:
It would be good to touch on how we gossip transactions. Are we gossiping based off of priority and/or sender?

Reply:
Yup added some comments about this in the discussion: https://github.com/tendermint/tendermint/discussions/6295#discussioncomment-642545

Reply:
Yes, I'll add a section for tx gossiping.